### PR TITLE
webgpu: Implement proper async pipeline creation and GPUPipelineError

### LINKED ISSUE #31591: Implement WebGPU async pipeline creation
Spec: https://gpuweb.github.io/gpuweb/#async-pipeline-creation
Stubs: https://github.com/servo/servo/pull/31194

How to do it: 

1. Store the promise on the device([here](https://github.com/servo/servo/pull/31194/files#diff-57b1e4d9cc2141b425d998d66541f5441c3e1ce4955c4713462e02132ba295e3R746)). Perhaps using a map of compute pipeline id to promise.  
2. Resolve the promise in [handle_server_msg](https://github.com/servo/servo/blob/3a5ca785d3ecc1fd6cb5a519cf1a91ac61e15c8c/components/script/dom/gpudevice.rs#L177)

### PR DESCRIPTION
**Pull request by sagudev**
~~This will require some changes in CTS: https://github.com/gpuweb/cts/issues/3809~~ DONE

Reviewable per commit.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #31591
- [x] There are tests for these changes in CTS

### CODE REVIEW

**Comment by sagudev** [comment]
try run: https://github.com/sagudev/servo/actions/runs/9962307544/job/27527741459

**Comment by sagudev** [comment]
NOTE-TO-SELF: `webgpu:api,validation,getBindGroupLayout:index_range,auto_layout:*` also fails on firefox, open issue on wgpu.

File: components/script/dom/gpurenderpipeline.rs [hunk 1]
```diff
@@ -86,13 +78,28 @@ impl GPURenderPipelineMethods for GPURenderPipeline {
...
+            .channel
+            .0
+            .send(WebGPURequest::RenderGetBindGroupLayout {
+                device_id: self.device.id().0,
+                pipeline_id: self.render_pipeline.0,
+                index,
+                id,
+            })
+        {
+            warn!("Failed to send WebGPURequest::ComputeGetBindGroupLayout {e:?}");
```
Comment (jdm) [inline components/script/dom/gpurenderpipeline.rs:96]: s/Compute/Render/

File: components/webgpu/ipc_messages/recv.rs [hunk 2]
```diff
@@ -103,6 +103,8 @@ pub enum WebGPURequest {
         compute_pipeline_id: id::ComputePipelineId,
         descriptor: ComputePipelineDescriptor<'static>,
         implicit_ids: Option<(id::PipelineLayoutId, Vec<id::BindGroupLayoutId>)>,
+        /// present only on ASYNC versions
+        sender: Option<IpcSender<WebGPUResponse>>,
```
Comment (jdm) [inline components/webgpu/ipc_messages/recv.rs:107]: Should we call this async_sender, in that case?

File: components/webgpu/wgpu_thread.rs [hunk 3]
```diff
@@ -448,7 +469,25 @@ impl WGPU {
                                 Some(render_pipeline_id),
                                 implicit)
                             );
-                            self.maybe_dispatch_wgpu_error(device_id, error);
+
+                            if let Some(sender) = sender {
```
Comment (jdm) [inline components/webgpu/wgpu_thread.rs:473]: It would be nice to avoid the code duplication here.

File: components/script/dom/gpudevice.rs [hunk 4]
```diff
@@ -214,6 +220,139 @@ impl GPUDevice {
...
+                }),
+                multisample: wgt::MultisampleState {
+                    count: descriptor.multisample.count,
+                    mask: descriptor.multisample.mask as u64,
+                    alpha_to_coverage_enabled: descriptor.multisample.alphaToCoverageEnabled,
+                },
+                multiview: None,
+            })
+        } else {
+            /*self.dispatch_error(webgpu::Error::Validation(String::from(
```
Comment (jdm) [inline components/script/dom/gpudevice.rs:348]: Can we either convert this into a clearer TODO comment or remove the commented code?

File: components/script/dom/gpudevice.rs [hunk 5]
```diff
@@ -214,6 +220,139 @@ impl GPUDevice {
...
+        descriptor: &GPURenderPipelineDescriptor,
+    ) -> (
+        Option<(PipelineLayoutId, Vec<BindGroupLayoutId>)>,
+        Option<RenderPipelineDescriptor<'c>>,
+    ) {
+        let mut valid = true;
+
+        let (layout, implicit_ids, _) = self.get_pipeline_layout_data(&descriptor.parent.layout);
+
+        let desc = if valid {
```
Comment (jdm) [inline components/script/dom/gpudevice.rs:234]: Valid is always true here. What is the intention behind this code?

File: components/script/dom/gpudevice.rs [hunk 6]
```diff
@@ -214,6 +220,139 @@ impl GPUDevice {
         }
     }
 
+    fn parse_render_pipeline<'c>(
+        &self,
+        descriptor: &GPURenderPipelineDescriptor,
+    ) -> (
+        Option<(PipelineLayoutId, Vec<BindGroupLayoutId>)>,
+        Option<RenderPipelineDescriptor<'c>>,
```
Comment (jdm) [inline components/script/dom/gpudevice.rs:228]: `c is not related to any method argument. Is that intentional? What is the risk if the output lifetime is completely made up?

**Review by jdm — APPROVED** [review]

File: components/webgpu/wgpu_thread.rs (hunk 3 above)
Comment (sagudev) [reply components/webgpu/wgpu_thread.rs:473]: You mean duplication between render/compute pipeline? They differentiate in `CreateRenderPipelineError` on `CreateComputePipelineError` and in responses they are sending, and I think dedup would require too much complexity (generics) for not enough profit (happens only twice).

Comment (jdm) [reply components/webgpu/wgpu_thread.rs:473]: 👍

File: components/script/dom/gpudevice.rs (hunk 6 above)
Comment (sagudev) [reply components/script/dom/gpudevice.rs:228]: Yes, it's intentional to make it compile. Output lifetime is automatically determined by compiler (to be 'static, but the compiler cannot determine that from function signature), so rewrote it to static lifetime to make it more obvious.

File: components/script/dom/gpudevice.rs (hunk 5 above)
Comment (sagudev) [reply components/script/dom/gpudevice.rs:234]: Ah, I planned to use https://docs.rs/wgpu/latest/wgpu/struct.ColorWrites.html#method.from_bits_retain to use validness check in wgpu-core, but I forgot.

Comment (sagudev) [reply components/script/dom/gpudevice.rs]: And it makes `webgpu:api,validation,render_pipeline,fragment_state:targets_write_mask:*` work.

**Review by jdm — APPROVED** [review]

**Pull request merged**